src/azure-cli/azure/cli/command_modules/role/custom.py

```diff
@@ -1745,17 +1745,25 @@ def _gen_guid():
 def _random_password(length):
     import random
     import string
-    random_source = string.ascii_letters + string.digits + string.punctuation
-    password = random.choice(string.ascii_lowercase)
-    password += random.choice(string.ascii_uppercase)
-    password += random.choice(string.digits)
-    password += random.choice(string.punctuation)
+    safe_punctuation = '@%_-+=:,.(){}[]<>'
```

Comment:
Thank you very much for submitting the PR. 

I don't think `<>()` are safe when they are not quoted. (Some users new to bash may forget to do that.) `<>` can be parsed as `stdin` `stdout` redirection, and `()` can be parsed as [subshell execution](https://www.gnu.org/software/bash/manual/html_node/Command-Grouping.html):

```
$ echo a>b
# Output is redirected

$ echo (ab)
-bash: syntax error near unexpected token `ab'
```

`%` is also not safe in a Windows Command Prompt: https://ss64.com/nt/syntax-percent.html

We will sync internally with the security team first and try to find out the best solution. 

Possible solutions are:

1. The first character must be a letter or number, then the rest characters can only use punctuation in `-_.~`
2. Show a warning when generating the password, something like "when using the password, it must be quoted to avoid shell interpretation"

Reply:
Microsoft Graph API generates the password by itself, instead of letting the user generate one https://github.com/Azure/azure-cli/issues/12561#issuecomment-605752672:

```
$ az rest -m "POST" -u https://graph.microsoft.com/v1.0/applications/b4e4d2ab-e2cb-45d5-a31a-98eb3f364001/addPassword --headers "Content-Type=application/json" -b '{"passwordCredential":{"displayName":"Password friendly name"}}'
{
  "@odata.context": "https://graph.microsoft.com/v1.0/$metadata#microsoft.graph.passwordCredential",
  "customKeyIdentifier": null,
  "displayName": "Password friendly name",
  "endDateTime": "2022-05-25T03:39:02.2389196Z",
  "hint": "4cc",
  "keyId": "b50bcc87-0f02-4aaf-b5c0-a871189871cb",
  "secretText": "4ccP5-~jc_fWK_4AJ-pzn.y2iNy~51tGnU",
  "startDateTime": "2020-05-25T03:39:02.2389196Z"
}
```

It only uses punctuation marks from `-_.~`. Some more examples:

```
4ccP5-~jc_fWK_4AJ-pzn.y2iNy~51tGnU
Pn~2983OPGQfi0.T.Jaai~v3664xRreKZi
-tw1b2Y46gfnzw0-2~WtfsbTpKy4q9H~8S
--zz5zs._yz-zSC5P~QcR23MNku3mQ2TVg
k-39_0y_83qmHgm.oMkypb._F7T7.5-Jbz
vaeV.4vDtu1HW~e-Fn-O.PM57leL1DqDa6
b2.vx-7Q9~Kn~Sc2Ii5dGnbEhPBj-1-PS7
~y4_UnMLgnnBu5.aXAQj-0j0X2p1-UEk_n
```

Reply:
If the password starts with dash `-`, Python's `argparse` will treat it as a parameter, like `--name`, instead of an argument. See https://github.com/Azure/azure-cli/issues/10692. 

For now, even if we only use letter or number as the first character, this issue will emerge after we migrate to Microsoft Graph (https://github.com/Azure/azure-cli/issues/12946).